### PR TITLE
[build] Support bzlmod for building and installing

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Towards #20731. See https://bazel.build/external/migration and https://bazel.build/external/overview#bzlmod for background.  As of the Bazel 9 release (late 2025), Bazel's (already deprecated) support for repository rules will be removed entirely, so Drake needs to port to bzlmod over the next ~12 months.  This is the first step -- adding a `MODULE.bazel` file and some related tweaks.  Future work involves converting our externals from repository rules to modules.

Note that this commit does not pass all tests when bzlmod is enabled; doing so requires enhancements to our runfiles handling (#22269).

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)

### CODE REVIEW

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 14 of 21 files at r1, 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot linux-jammy-unprovisioned-gcc-wheel-experimental-release please

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 10 of 21 files at r1, 8 of 10 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: needs platform reviewer assigned, needs at least two assigned reviewers

**Comment by jwnimmer-tri** [comment]
@drake-jenkins-bot mac-arm-sonoma-unprovisioned-clang-wheel-experimental-release please

**Comment by xuchenhan-tri** [comment]
+@xuchenhan-tri

**Review by xuchenhan-tri — COMMENTED** [review]
checkpoint. I'll come back on Monday for the test.

> Reviewed 10 of 21 files at r1, 7 of 10 files at r2, 3 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-:-ODTh0iUAD-s7Nyggskq:b7j0ukn)* status: 6 unresolved discussions, LGTM missing from assignee xuchenhan-tri(platform), needs at least two assigned reviewers

___
*[`-- commits` line 15 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTgp7S4sChmd57nobL:-ODTgp7S4sChmd57nobM:b-skuz4z):*
BTW, I can't seem to observe the effect of this.

I tried running `bazel run //:install` with and without this commit and melded the results.



_Code quote:_
```
- Adjust install rule so installed paths are "drake" not "_main".
```
___
*[`CMakeLists.txt` line 548 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTf36xAEvWZWEyTHQZ:-ODTf36xAEvWZWEyTHQ_:b-wln4ru) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/CMakeLists.txt#L548)):*
> ```CMake
> file(CREATE_LINK "${PROJECT_SOURCE_DIR}/.bazeliskrc" drake_build_cwd/.bazeliskrc SYMBOLIC)
> file(CREATE_LINK "${PROJECT_SOURCE_DIR}/MODULE.bazel" drake_build_cwd/MODULE.bazel SYMBOLIC)
> file(CREATE_LINK "${PROJECT_SOURCE_DIR}/WORKSPACE" drake_build_cwd/WORKSPACE SYMBOLIC)
> ```

BTW, why do we still need the WORKSPACE file now that bzlmod is enabled for CMake?
___
*[`tools/workspace/default.bzl` line 112 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODT_av51N8b15umn6-5:-ODT_av51N8b15umn6-6:b-e2jzo9) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/workspace/default.bzl#L112)):*
> ```Starlark
> load("//tools/workspace/zlib:repository.bzl", "zlib_repository")
> 
> # This is the list of modules that our MODULE.bazel already incorporates.
> ```

nit this is not exactly accurate. There's also rules_java that's module only.



_Suggestion:_
```Starlark
This is the list of modules to be excluded when bzlmod is enabled because our MODULE.bazel already incorporates them.
```
___
*[`tools/py_toolchain/BUILD.bazel` line 26 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTaOCL1b74iXliYvTN:-ODTaOCL1b74iXliYvTO:bstz7sz) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/py_toolchain/BUILD.bazel#L26)):*
> ```Starlark
> )
> 
> py_exec_tools_toolchain(
> ```

BTW, TBH, I still haven't quite understand the toolchain change. Can you explain again?
___
*[`tools/workspace/rules_python/repository.bzl` line 29 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTbH2r9Psa7tk3p9Ix:-ODTbH2r9Psa7tk3p9Iy:b-8u8vlc) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/workspace/rules_python/repository.bzl#L29)):*
> ```Starlark
>         name,
>         mirrors = None,
>         _constants_only = False):
> ```

BTW, it took me a while (and the context of a previous PR) to understand this argument (and why it's named as such). It needs some documentation.
___
*[`MODULE.bazel` line 18 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTSGmq15VMRTtoT6tk:-ODTSGmq15VMRTtoT6tl:b-z5t9em) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/MODULE.bazel#L18)):*
> ```Starlark
> bazel_dep(name = "rules_java", version = "8.6.1")
> bazel_dep(name = "rules_license", version = "1.0.0")
> bazel_dep(name = "rules_python", version = "0.40.0")
> ```

BTW, is this an exhaustive list of all of our external dependencies that are registered in the Bazel Central Registry?

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 4 of 4 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: 5 unresolved discussions, LGTM missing from assignee xuchenhan-tri(platform), needs at least two assigned reviewers

___
*[`-- commits` line 15 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTgp7S4sChmd57nobL:-ODTiWfY8g4OI9qivP6f:byc8vgc):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, I can't seem to observe the effect of this.

I tried running `bazel run //:install` with and without this commit and melded the results.
</blockquote></details>

Yes, this commit _preserves_ the current behavior.

If you revert the changes to the second hunk of the `drake/BUILD.bazel` patch (the `install(name = "install_files", ...)` thing) while keeping the rest of the PR the same (i.e., with `--enable_bzlmod=true`), then I believe you would see the install tree have a couple of files in `_main` instead of `drake`.

It's possible that only a `cmake ../drake && make install` shows the erroneous behavior (running the install from CMake vs Bazel) but IIRC both were broken.
___
*[`CMakeLists.txt` line 548 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTf36xAEvWZWEyTHQZ:-ODTivLF3SDJ1kxkqYH_:b-d3zg8o) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/CMakeLists.txt#L548)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, why do we still need the WORKSPACE file now that bzlmod is enabled for CMake?
</blockquote></details>

Without a `WORKSPACE` file to start with, it seems that Bazel (in certain cases?) does not even look for the `WORKSPACE.bzlmod` file.  If we remove this line, a subset of the CI builds will fail (IIRC only one of the pre-merge jobs, but several of the post-merge ones).  In other words, it serves as a sentinel, but never gets run.
___
*[`MODULE.bazel` line 18 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTSGmq15VMRTtoT6tk:-ODTjbEW9M6metQmhgDN:b-ye5dhv) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/MODULE.bazel#L18)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, is this an exhaustive list of all of our external dependencies that are registered in the Bazel Central Registry?
</blockquote></details>

Not yet.  After the next PR (#22269), I believe we'll have all of the _starlark_  dependencies from BCR (i.e., rules and toolchains) listed in `MODULE.bazel`.

The general idea of this whole train is that we will incrementally keep adding more and more dependencies into `MODULE.bazel` (so that `add_default_workspace(bzlmod = True)` does less and less work -- by growing the list of `REPOS_ALREADY_PROVIDED_BY_BAZEL_MODULES`) until we can delete `WORKSPACE.bzlmod` entirely, at which point we are purely either 100% workspace or 100% bzlmod.  At that point (or maybe a little sooner), we can deprecate our support for workspace for eventual removal entirely.

I could put some of that backstory into either the issue, or a comment in this file, if that helps?
___
*[`tools/py_toolchain/BUILD.bazel` line 26 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTaOCL1b74iXliYvTN:-ODTndCK0Ud1nHBp90ze:bm4kgr) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/py_toolchain/BUILD.bazel#L26)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, TBH, I still haven't quite understand the toolchain change. Can you explain again?
</blockquote></details>

This one is slightly different than https://github.com/RobotLocomotion/drake-external-examples/pull/344.

Over there, we disabled the exec interpreter.  Here, this is the magic incantation for "make the exec interpreter be the same as the normal interpreter".  (This stanza might have worked over there, too, it just took me a while to find this one so I never tried it in DEE.)

The exec toolchain is described a bit here: https://rules-python.readthedocs.io/en/latest/toolchains.html#exec-tools-toolchain-type

Basically, it's the python interpreter which is used internal to the build, as opposed to the normal python interpreter that we're targeting (e.g., the one used to run/test pydrake and its tutorials).
___
*[`tools/workspace/default.bzl` line 112 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODT_av51N8b15umn6-5:-ODTkq0BEizX1E8PEVVR:bnx5n3i) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/workspace/default.bzl#L112)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

nit this is not exactly accurate. There's also rules_java that's module only.
</blockquote></details>

Actually, that's a good point.  If I adjust this list to exactly match `MODULE.bazel`, then the test will get a smidge simpler and the comment here doesn't need to dance around the laser beams.

Done.
___
*[`tools/workspace/rules_python/repository.bzl` line 29 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTbH2r9Psa7tk3p9Ix:-ODTnPkw9me4CpePl38x:bb6un1j) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/workspace/rules_python/repository.bzl#L29)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

BTW, it took me a while (and the context of a previous PR) to understand this argument (and why it's named as such). It needs some documentation.
</blockquote></details>

Done

**Comment by jwnimmer-tri** [comment]
+@ggould-tri for platform review per schedule (tomorrow), please.

**Review by xuchenhan-tri — COMMENTED** [review]
:lgtm: 

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-:-ODgv7O_BCwT68b13bYN:bg7igkv)* status: 1 unresolved discussion, LGTM missing from assignee ggould-tri(platform)

___
*[`-- commits` line 15 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTgp7S4sChmd57nobL:-ODgrTh15Qjp8EvjUmKS:b-402whg):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, this commit _preserves_ the current behavior.

If you revert the changes to the second hunk of the `drake/BUILD.bazel` patch (the `install(name = "install_files", ...)` thing) while keeping the rest of the PR the same (i.e., with `--enable_bzlmod=true`), then I believe you would see the install tree have a couple of files in `_main` instead of `drake`.

It's possible that only a `cmake ../drake && make install` shows the erroneous behavior (running the install from CMake vs Bazel) but IIRC both were broken.
</blockquote></details>

Ok, I can reproduce the error from bazel build. I don't understand why the diff fixes the error though... Care to explain (perhaps in the BUILD file as comments)?
___
*[`CMakeLists.txt` line 548 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTf36xAEvWZWEyTHQZ:-ODgl-XV02JQeglAc043:b-aoo0wt) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/CMakeLists.txt#L548)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Without a `WORKSPACE` file to start with, it seems that Bazel (in certain cases?) does not even look for the `WORKSPACE.bzlmod` file.  If we remove this line, a subset of the CI builds will fail (IIRC only one of the pre-merge jobs, but several of the post-merge ones).  In other words, it serves as a sentinel, but never gets run.
</blockquote></details>

Ok, interesting. Thanks!
___
*[`MODULE.bazel` line 18 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTSGmq15VMRTtoT6tk:-ODgkXuK15ODqPcVfMUO:b-jn1a2b) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/MODULE.bazel#L18)):*
> I could put some of that backstory into either the issue, or a comment in this file, if that helps?

I don't think that's necessary, but certainly go ahead and do that if you think it might help somebody.

The reason that I asked is because the list seems so much shorter than our full set of dependencies...
___
*[`tools/py_toolchain/BUILD.bazel` line 26 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTaOCL1b74iXliYvTN:-ODgr8vnAVBLK-CiGdG1:b-vesdcp) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/py_toolchain/BUILD.bazel#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This one is slightly different than https://github.com/RobotLocomotion/drake-external-examples/pull/344.

Over there, we disabled the exec interpreter.  Here, this is the magic incantation for "make the exec interpreter be the same as the normal interpreter".  (This stanza might have worked over there, too, it just took me a while to find this one so I never tried it in DEE.)

The exec toolchain is described a bit here: https://rules-python.readthedocs.io/en/latest/toolchains.html#exec-tools-toolchain-type

Basically, it's the python interpreter which is used internal to the build, as opposed to the normal python interpreter that we're targeting (e.g., the one used to run/test pydrake and its tutorials).
</blockquote></details>

I _think_ I understood. Maybe I'll come to appreciate this more fully another day.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: LGTM missing from assignee ggould-tri(platform)

___
*[`-- commits` line 15 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTgp7S4sChmd57nobL:-ODgwBSgANNhLnX7UMnu:b-6fhdpg):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

Ok, I can reproduce the error from bazel build. I don't understand why the diff fixes the error though... Care to explain (perhaps in the BUILD file as comments)?
</blockquote></details>

Done (new comment in BUILD.bazel).
___
*[`MODULE.bazel` line 18 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTSGmq15VMRTtoT6tk:-ODgwy_1CJxOKJSonqSn:bw4mste) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/MODULE.bazel#L18)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

> I could put some of that backstory into either the issue, or a comment in this file, if that helps?

I don't think that's necessary, but certainly go ahead and do that if you think it might help somebody.

The reason that I asked is because the list seems so much shorter than our full set of dependencies...
</blockquote></details>

Indeed, a TODO would be very helpful to future readers.  Added that now.
___
*[`tools/py_toolchain/BUILD.bazel` line 26 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODTaOCL1b74iXliYvTN:-ODgwHCS0Ua8RcbnmIXi:b-or6smk) ([raw file](https://github.com/RobotLocomotion/drake/blob/3c94c759e67026500cc32ad78887c2f733cbd9d8/tools/py_toolchain/BUILD.bazel#L26)):*
<details><summary><i>Previously, xuchenhan-tri wrote…</i></summary><blockquote>

I _think_ I understood. Maybe I'll come to appreciate this more fully another day.
</blockquote></details>

I probably should have also said -- most acutely, this exec interpreter is the interpreter used by the `mkdoc` tool to produce `documentation_pybind.h`; if it ends up being set incorrectly (i.e., if we don't set it here) then we can't even compile pydrake.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: LGTM missing from assignee ggould-tri(platform)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees ggould-tri(platform),xuchenhan-tri(platform)

___
*[`tools/skylark/drake_py.bzl` line 153 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/22271#-ODl6VdJCw0BTCqKz9Zw:-ODlCvyd55uk9U4f-uUT:be8dd86) ([raw file](https://github.com/RobotLocomotion/drake/blob/f4d6448ec410d55130866c1ea96b178519f71790/tools/skylark/drake_py.bzl#L153)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

minor:  I don't know what this means.  Is this a reference to the last item of https://rules-python.readthedocs.io/en/stable/precompiling.html ?

(It seems harmless, just mysterious)
</blockquote></details>

I added more commentary.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22271)* status: <span>:shipit:</span> complete! all discussions resolved, LGTM from assignees ggould-tri(platform),xuchenhan-tri(platform)
